File: src/python/scripts/analysis/s9c_qc_prep.R
#!/usr/bin/env Rscript
#
# Reads gene and output nRegion x nPuck matrix for genes specified in argument. 
# This script is used by s9c_qc_main.py
#
# Usage example:
#
#  Rscript src/python/scripts/analysis/s9c_qc_prep.R
#
# Created by Mukund on 2022-06-11

library(anndata)
Sys.setenv(RETICULATE_PYTHON = "/Users/mraj/opt/miniconda3/envs/Rv4/bin/python3")
library(reticulate)
library(Matrix)
library(Matrix.utils)



args = commandArgs(trailingOnly=TRUE)
print(args)
# 
# gene = 'Ndnf'
# gene = 'Gad1'
# ip_folder_labels <- "/Users/mraj/Desktop/work/data/mouse_atlas/data_v3_nissl_post_qc/s9_analysis/aggregated_labels"
# ip_folder_counts <- "/Users/mraj/Desktop/work/data/temp_data/2022-05-04/integrated_mats"
# r_op_folder_aggr <- "/Users/mraj/Desktop/work/data/mouse_atlas/data_v3_nissl_post_qc/s9_analysis/qc_mats/aggr_data"
# region_ids_str <- '1;2;3;4;5;6'
# start_pid <- 79
# end_pid <- 81
# nonzero <- FALSE
# genes <- 'Ndnf,Gad1'


# region_ids <- '1006;670;1086;1111;9;461;1089'

genes <- args[1]
ip_folder_labels <- args[2]
ip_folder_counts <- args[3]
r_op_folder_aggr <- args[4]
region_ids_str <- args[5]
start_pid <- as.integer(args[6])
end_pid <- as.integer(args[7])
nonzero <- args[8]=='True'


# br_layer_ids <- strsplit(region_ids,";")
# rids_lists <- vector(mode = "list", length = length(br_layer_ids[[1]]))
# ii=0
# for (brlstr in br_layer_ids[[1]]){
#   ii=ii+1
# 
#   brl_ids <- strsplit(brlstr[[1]], ',')
#   rids_lists[[ii]] <- vector(mode = "list", length = length(brl_ids[[1]]))
#   # brl_ids <- as.numeric(unlist(brl_ids))
#   # brl_ids <- unlist(brl_ids)
# 
#   jj=0
#   for (brl_id in brl_ids){
#         jj=jj+1
# 
#         rids_lists[[ii]][[jj]] <- brl_id
#   }
# }

genes_names <- strsplit(genes,',')[[1]]
region_ids <- strsplit(region_ids_str,";")[[1]]
gene_mats <- vector(mode = "list", length = length(genes_names))
gene_mats_raw<- vector(mode = "list", length = length(genes_names))
genes_idxs <- vector( "integer" , length(genes_names) )




pucks <- seq(start_pid,end_pid,2)

for (gene_i in seq_along(genes_names)){
  gene_mats[[gene_i]] <- matrix(0, nrow = length(region_ids), ncol = length(pucks))
  gene_mats_raw[[gene_i]] <- matrix(0, nrow = length(region_ids), ncol = length(pucks))
}

print(pucks)
Xmeans <- vector( "numeric" , length(pucks) )
Xsds <- vector( "numeric" , length(pucks) )
# loop over pucks
pid_idx <- 0
for (pid in pucks) {
  print(pid)
  pid_idx <- pid_idx + 1
  apid = pid
  if (pid==5 || pid==77 || pid==167)
    apid = pid - 2 ## adjusted pid todo: modify viewer to not require this adjustment

  # get data matrix of puck
  ip_counts_file = paste0(ip_folder_counts, "/ad_counts_", apid, ".h5ad")
  print(ip_counts_file)
  ad <- read_h5ad(ip_counts_file)
  

  if (pid_idx==1){  # populate gene_inds only one time
    for(i in seq_along(genes_names)){
      gene_search_str <- paste0("^",{genes_names[i]},"$")
      genes_idxs[i] <- grep(gene_search_str, ad$obs_names)
    }
  }
  X <- t(ad$X)
  Xmeans[pid_idx] <- mean(X@x)
  Xsds[pid_idx] <- sd(X@x)
  # Xsd <- sd(X)
  # print(dim(X))
  # print(length(ad$obs_names))
  # filter selected gene
  # gene_search_str <- paste0("^", gene, "$")
  # gene_idx <- ad$obs_names==gene
  #X_filt <- X[, gene_idx]
  X_filt <- X[, genes_idxs]
  #totals_umis <- rowSums(X) # total umis in puck split by gene
  #total_umis_gene <- totals_umis[gene_idx]
  #print(length(totals_umis))
  print(c('len X_filt', length(X_filt)))
  
  

  # get labels factor
  ip_labels_file = paste0(ip_folder_labels, '/agg_labels_',sprintf("%03d", apid),'.csv')
  print(ip_labels_file)
  labels <- factor(read.csv(file=ip_labels_file, header=FALSE)$V1)
  
  # aggregate over regions
  b<-aggregate.Matrix(X_filt,labels, FUN=sum()) # regions x chosen_gene(s)
  print("bazin")
  print(b)
  
  #cmat<-aggregate.Matrix(X,labels, FUN=sum()) # region x all_gene
  #print(b)
  
  #gene_mat <- matrix(0, nrow = length(region_ids), ncol = length(pucks))
  
  
  print('entering loop')
  # populate heatmap column for this puck by iterating over required genes and regions
  for (gene_i in seq_along(genes_idxs)){
    for (region_i in seq_along(region_ids)){
      
      if (region_ids[region_i] %in% row.names(b)){
        cnt <- b[region_ids[region_i],][genes_names[gene_i]][[1]]
      }
      else{
        cnt <- 0
      }
      #print(b)
      #print(cnt[genes_names[gene_i]][[1]])
     
      gene_mats_raw[[gene_i]][region_i, pid_idx] <- cnt
      gene_mats[[gene_i]][region_i, pid_idx] <- (gene_mats_raw[[gene_i]][region_i, pid_idx] - Xmeans[pid_idx])/Xsds[pid_idx]
    }
    #gene_mats[[gene_i]][,pid_idx] <- gene_mat[,pid_idx]
  }
  # print(gene_mats_raw)
  # if(0)
  # for (ii in seq(length(rids_lists))){ # looping over region
  #  
  #   sum <- 0;
  #   for (jj in seq(length(rids_lists[[ii]]))){ ## looping over sub region
  # 
  #     # get gene count in sub region
  #     if (any(levels(labels)==rids_lists[ii][jj][[1]][[1]])){
  #       total_region_umis <- sum(cmat[rids_lists[ii][jj][[1]][[1]],])
  #       cnt <- b[rids_lists[ii][jj][[1]][[1]], , drop = FALSE]@x
  #       if (length(cnt)==1){
  #         # update count in mat
  #         if (nonzero){
  #           gene_mat[ii, pid_idx] <- gene_mat[ii, pid_idx]+1
  #         }else{
  #           gene_mat[ii, pid_idx] <- gene_mat[ii, pid_idx]+cnt/total_region_umis
  #           #gene_mat[ii, pid_idx] <- gene_mat[ii, pid_idx]+cnt
  #           #print(c(ii,jj,rids_lists[[ii]][[jj]][[1]][[1]]))
  #         }
  #         
  #       }
  #      
  #     }
  #   }
  # }
  
  # normalize by column sum to address puck specific effects
  # gene_mat[,pid_idx] <- gene_mat[,pid_idx]/sum(cmat)
  
  gc()
  
}



# (gene_mat - mean(gene_mat)/sd(gene_mat))
# write out output matrix csvsd
for (gene_i in seq_along(genes_names)){
  
  op_raw_file_csv <- paste0(r_op_folder_aggr,'/',genes_names[gene_i],'_',start_pid,'_',end_pid,'_raw.csv')

  write.table(gene_mats_raw[[gene_i]], file=op_raw_file_csv, row.names=FALSE, col.name=FALSE, sep=',')
  
  op_file_csv <- paste0(r_op_folder_aggr,'/',genes_names[gene_i],'_',start_pid,'_',end_pid,'.csv')
  print(op_file_csv)
  write.table(gene_mats[[gene_i]], file=op_file_csv, row.names=FALSE, col.name=FALSE, sep=',')
  
}

#write.table(gene_mat, file=r_op_filename, row.names=FALSE, col.name=FALSE, sep=',')


# matrix(, nrow = 2, ncol = 2)
# any(levels(labels)=="1016")
# b['1016', , drop = FALSE]@x
# write.table(mm, file='temp.csv', row.names=FALSE, col.name=FALSE, sep=',')
# 
# https://stat.ethz.ch/pipermail/r-help/2011-April/275773.html
# r_op_filename = '/Users/mraj/Desktop/work/data/mouse_atlas/temp/Ndnf-cnt.csv'
# gene_mat <- as.matrix(read.table(file=r_op_filename, sep=','))
# colnames(gene_mat) = pucks
# rownames(gene_mat) = c('L1', 'L2/3','L4', 'L5', 'L6', 'L7', 'Hippo Formation' )
# heatmap(gene_mat, Colv = NA, Rowv = NA, scale="column")

# #Initial Heatmap 
# install.packages("tidyverse")
# library(tidyverse)
# gene_mat %>% 
#   as.data.frame() %>%
#   rownames_to_column("f_id") %>%
#   pivot_longer(-c(f_id), names_to = "samples", values_to = "counts") %>%
#   mutate(samples= fct_relevel(samples,colnames(dat))) %>%
#   ggplot(aes(x=samples, y=f_id, fill=counts)) + 
#   geom_raster() + 
#   scale_fill_viridis_c()
# convert gene_mat values to z scores and write out

File: src/python/scripts/analysis/s9c_qc_prep.py
"""
Script for generating identifing and storing bead indices from specific regions

Usage:

python s9c_qc_prep.py \
    inp: data root path
    inp: path to bead_ccf_labels_allbds with bead metadata in csv format
    inp: path to bead_ccf_coords_allbds with bead coords in chuck space in csv format
    inp: input to integrated_mats folder with processed gene counts in annodata h5ad format
    out: output dir to write qc_mats

Usage example:

python src/python/scripts/analysis/s9c_qc_prep.py \
    /Users/mraj/Desktop/work/data/mouse_atlas \
    /data_v3_nissl_post_qc/s7_annotations/bead_ccf_labels_allbds \
    /data_v3_nissl_post_qc/s3_registered_ss/chuck_img_coords_allbds \
    /Users/mraj/Desktop/work/data/temp_data/2022-05-04/integrated_mats \
    /data_v3_nissl_post_qc/s9_analysis/qc_mats

Created by Mukund on 2022-06-09
"""
import sys
from produtils import dprint
import csv
import anndata as ann
import numpy as np
from scipy.sparse import csr_matrix, hstack
import json

from allensdk.core.reference_space_cache import ReferenceSpaceCache

reference_space_key = 'annotation/ccf_2017'
resolution = 25
rspc = ReferenceSpaceCache(resolution, reference_space_key, manifest='manifest.json')
# ID 1 is the adult mouse structure graph
tree = rspc.get_structure_tree(structure_graph_id=1) 

ip_data_root = sys.argv[1]
ip_folder_labels = ip_data_root+sys.argv[2]
ip_folder_chuck_coords = ip_data_root+sys.argv[3]
ip_folder_counts = sys.argv[4]
op_folder = ip_data_root+sys.argv[5]

# regional_inds = {
#     "Primary somatosensory area, trunk, layer 1":[],
#     "Primary somatosensory area, trunk, layer 2/3":[],
#     "Primary somatosensory area, trunk, layer 4":[],
#     "Primary somatosensory area, trunk, layer 5":[],
#     "Primary somatosensory area, trunk, layer 6a":[],
#     "Primary somatosensory area, trunk, layer 6b":[],
#     "Hippocampal formation":[]
# }

region_name_to_id = {
    'Primary somatosensory area, trunk, layer 1':-1,
    'Primary somatosensory area, trunk, layer 2/3':-1,
    'Primary somatosensory area, trunk, layer 4':-1,
    'Primary somatosensory area, trunk, layer 5':-1,
    'Primary somatosensory area, trunk, layer 6a':-1,
    'Primary somatosensory area, trunk, layer 6b':-1,
    'Hippocampal formation':-1
}
for rname in region_name_to_id:
    region_name_to_id[rname] = tree.get_structures_by_name([rname])[0]['id']
dprint(region_name_to_id)

genes_list = ['Dcn', 'Ndnf']

pucks = list(range(1, 208, 2))
data_dicts = []
dprint(pucks)
for i in range(len(genes_list)):
    data = {"x": list(range(1, len(pucks)+1)), "y": [1,2], "z":[[],[]]}
    data_dicts.append(data)
dprint(data_dicts)
# exit(0)
# get number of regions
for pid in pucks:

    dprint("starting pid ", pid)
    apid = pid
    if (pid==5 or pid==77 or pid==167):
        apid = pid - 2 ## adjusted pid todo: modify viewer to not require this adjustment

    # ip_coords_file  = f'{ip_folder_counts}/ad_coords_{str(apid)}.h5ad'
    ip_counts_file  = f'{ip_folder_counts}/ad_counts_{str(apid)}.h5ad'

    # get region names/labels
    nis_id_str = str(apid).zfill(3)
    labels_csv_file = f'{ip_folder_labels}/allen_anno_data_{nis_id_str}.csv'
    region_names = []
    out_tissue = []
    with open(labels_csv_file, newline='\n') as csvfile:
        reader = csv.reader(csvfile)
        for row in reader:
            region_names.append(row[4])
            out_tissue.append(row[8])

    counts = ann.read_h5ad(ip_counts_file)
    counts_X = csr_matrix(counts.X).transpose()
    genes = list(counts.obs_names)
    gene_count_total = counts_X.sum(axis=1)
    # dprint(counts_X.get_shape())
    dprint('gene_count_total shape:', np.shape(gene_count_total))

    nis_id_str = str(apid).zfill(3)
    # get chuck space img coords
    chuck_sp_img_coords_file = f'{ip_folder_chuck_coords}/chuck_sp_img_coords_{nis_id_str}.csv'
    dprint('chuck_sp_img_coords file ', chuck_sp_img_coords_file)
    chuck_sp_img_coords = []
    with open(chuck_sp_img_coords_file, newline='\n') as csvfile:
        reader = csv.reader(csvfile)
        for row in reader:
            chuck_sp_img_coords.append([int(row[0]), int(row[1])])
    dprint(len(chuck_sp_img_coords))
    max_len = len(chuck_sp_img_coords)

    regional_inds = {
        "Primary somatosensory area, trunk, layer 1":[],
        "Primary somatosensory area, trunk, layer 2/3":[],
        "Primary somatosensory area, trunk, layer 4":[],
        "Primary somatosensory area, trunk, layer 5":[],
        "Primary somatosensory area, trunk, layer 6a":[],
        "Primary somatosensory area, trunk, layer 6b":[],
        "Hippocampal formation":[]
    }

    for idx in range(len(chuck_sp_img_coords)):
        leaf_region_name = region_names[idx]
        assert(idx<max_len)
        if (out_tissue[idx]=='F'):
            # if (idx>170950):
            #     dprint(idx)
            leaf_id = tree.get_structures_by_name([leaf_region_name])[0]['id']
            # dprint(leaf_region_name, leaf_id)
            # if tree.structure_descends_from(leaf_id, 1006):
            if leaf_id==1006:
                # Primary somatosensory area, trunk, layer 1
                regional_inds['Primary somatosensory area, trunk, layer 1'].append(idx)
            # elif tree.structure_descends_from(leaf_id, 670):
            elif leaf_id==670:
                # Primary somatosensory area, trunk, layer 2/3
                regional_inds['Primary somatosensory area, trunk, layer 2/3'].append(idx)
            # elif tree.structure_descends_from(leaf_id, 1086):
            elif leaf_id==1086:
                # Primary somatosensory area, trunk, layer 4
                regional_inds['Primary somatosensory area, trunk, layer 4'].append(idx)
            # elif tree.structure_descends_from(leaf_id, 1111):
            elif leaf_id==1111:
                # Primary somatosensory area, trunk, layer 5
                regional_inds['Primary somatosensory area, trunk, layer 5'].append(idx)
            # elif tree.structure_descends_from(leaf_id, 9):
            elif leaf_id==9:
                # Primary somatosensory area, trunk, layer 6a
                regional_inds['Primary somatosensory area, trunk, layer 6a'].append(idx)
            # elif tree.structure_descends_from(leaf_id, 461):
            elif leaf_id==461:
                # Primary somatosensory area, trunk, layer 6b
                regional_inds['Primary somatosensory area, trunk, layer 6b'].append(idx)
            elif tree.structure_descends_from(leaf_id, 1089):
                # Hippocampal formation
                regional_inds['Hippocampal formation'].append(idx)

            # if tree.structure_descends_from(str_id, cerebral_cortex_id):
            #     gene_count_cortex += spec_gene_cnts_dense[idx]
            #     total_count_cortex += gene_count_total[idx][0,0]
            # if tree.structure_descends_from(str_id, hippo_id):
            #     gene_count_hippo += spec_gene_cnts_dense[idx]
            #     total_count_hippo += gene_count_total[idx][0,0]
    # dprint(regional_inds)
    with open(f'{op_folder}/regional_inds_{pid}.json', "w") as outfile:
        json.dump(regional_inds, outfile)
